### ISSUE TITLE
Proposition: subcommands

### ISSUE THREAD

**Issue by EmmChriss** [mawww/kakoune#2429]
As I understand, and often experience, in my opinion, kakoune's scripting language would need a some additional level of structure; thus I propose the command variant `def-command [<switches>] <scope> <name> [<command>:<name>]`, which would practically make the listed commands its subcommands - perhaps under altered names

To make my idea clearer, I'd like to offer an example
```
def lsp-start %{ ... }
def lsp-hover %{ ... }
def lsp-defition %{ ... }
def lsp-references %{ ... }
```
This is a pseudo-excerpt from ul's kak-lsp plugin.

In this example, one would have to type the 'lsp-' part in the command line to get autocompletion for the commands. Everything's fine until `lsp-inline-diagnostic-enable`, `lsp-inline-diagnostic-disable`, and the alike enter the picture; and the intellisense becomes a jarbled mess with no structure or hierarchy.
However, if the definition of subcommands would be possible, a nice hierarchy of commands would also be, and typing 'lsp' would display a single command, that shows the possible subcommands, as well as a piece of documentation perhaps.

**Comment by maximbaz** [comment]
Is the intention purely to reduce the number of candidates in intellisense? I would argue the current behavior is preferable, for example I can type `:ref` and immediately discover `lsp-references`, going through `lsp` command to find `references` subcommand would be slower.

Having a `lsp` user mode with all available commands could be useful though, cc @ul 😉

**Comment by EmmChriss** [comment]
Didn't even consider that side of the situation...
The -enable, and -disable commands definitely need a new mechanism though

**Comment by mawww** [comment]
I agree there is some cluttering going in with commands, that said I think we already have (almost) all the necessary tools to make those commands:

```
def lsp -params 1.. -shell-candidates %{for cmd in start hovel definition...; do echo $cmd; done} %{ eval "lsp-%arg{1}" }
```

**Comment by maximbaz** [comment]
mnemonics choice looks good to me

**Comment by EmmChriss** [comment]
Looks really nice, this is just about what I wanted, thanks a lot @ul

**Comment by ul** [comment]
What do you think about the following solution for `*-{enable,disable}` mess: just define flags as options instead of commands and use `*SetOption` hooks to react to their change?

**Comment by Emmk** [comment]
That sounds a lot less messy, than what we have now; I definitely think it's a good idea

**Comment by Delapouite** [comment]
A similar "sub command" proposal with "masks": https://github.com/mawww/kakoune/issues/1081
